README.md
# tru.ID SDK for Mobile Web Applications

[![License][license-image]][license-url]

The Web SDK for tru.ID providing a helper function for requesting the `check_url` for [PhoneCheck](https://tru.id/docs/phone-check) and [SubscriberCheck](https://tru.id/docs/subscriber-check).

## Installation

### Via jsDelivr CDN

```html
<script src="https://cdn.jsdelivr.net/npm/@tru_id/tru-sdk-web/dist/tru-id-sdk.umd.js"></script>
```

### Via NPM

```sh
$ npm install @tru_id/tru-sdk-web
```

## Usage

### Via jsDelivr CDN

#### Reachability Check

```html
<script src="https://cdn.jsdelivr.net/npm/@tru_id/tru-sdk-web/dist/tru-id-sdk.umd.js"></script>
<script>
  tru.ID.getReachability(url)
</script>
```

#### PhoneCheck

When installed via a CDN a `tru.ID` global is installed.

```html
<script src="https://cdn.jsdelivr.net/npm/@tru_id/tru-sdk-web/dist/tru-id-sdk.umd.js"></script>
<script>
  tru.ID.openCheckUrl(url, config)
</script>
```

### Via NPM

#### Reachability Check

```js
import truID from '@tru_id/tru-sdk-web'

truID.getReachability(url)
```

#### PhoneCheck

When installed via NPM the imported object exposes the `openCheckUrl` function.

```js
import truID from '@tru_id/tru-sdk-web'

truID.openCheckUrl(checkUrl, config)
```

### `openCheckUrl` Config

The `openCheckUrl` function takes an optional `config` Object argument:

```js
truID.openCheckUrl(url, config)
```

The configuration options are:

```js
{
  // whether debug information will be logged to the console.
  // Defaults to `false`
  debug: boolean,

  // "image" - a zero pixel image is dynamically added to the DOM for the check_url request
  // "window" - `window.open` is called to open the check_url in a new window
  // Defaults to "image".
  checkMethod: "image" | "window",

  // If `checkMethod` was set to `window` identifies the number of
  // milliseconds after which the opened window will be closed.
  // Defaults to 3000.
  windowCloseTimeout: Number,

  // It will run the device coverage check to determine
  // if the device is on a mobile IP
  // Defaults to true
  checkDeviceCoverage: boolean,

  // Depending on what PhoneCheck API version the check is created with.
  // Defaults to "v0.2"
  version: "v0.1" | "v0.2",
}
```

## Local development

You can run `npm run dev` that will open rollup with watch mode that will re-compile the SDK after every change.

In another terminal you can run `npm run serve` to open a test web page where you can test real phone checks if you have the node server running or you can simply check a PhoneCheck `check_url`.

## Releasing

1. Bump the version if required
2. Run `npm run changelog` to auto-update the `CHANGELOG.md` and manually edit to finesse
3. Commit the changes read for release: `git commit -m 'chore(release): v{version}`
4. Tag the release `git tag v{version}`
5. Push the tags `git push --follow-tags origin main`
6. Build the project with `npm run build`
6. Publish to NPM `npm publish --access public`

## License

[MIT](LICENSE)

## Meta

Distributed under the MIT license. See ``LICENSE`` for more information.

[https://github.com/tru-ID](https://github.com/tru-ID)

[license-image]: https://img.shields.io/badge/License-MIT-blue.svg
[license-url]: LICENSE


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.11](https://github.com/tru-ID/tru-sdk-web/compare/v0.0.10...v0.0.11) (2022-07-22)

* Added Check version v0.2 functionality.

### [0.0.10](https://github.com/tru-ID/tru-sdk-web/compare/v0.0.8...v0.0.10) (2022-03-24)

* Fixed typo in usage example in `README`, incorrectly referring to `url` as `checkUrl`.

### [0.0.9](https://github.com/tru-ID/tru-sdk-web/compare/v0.0.8...v0.0.9) (2022-03-24)

* Added improvements to structure of `README`, and moved example from `README` to `example` directory

### [0.0.8](https://github.com/tru-ID/tru-sdk-web/compare/v0.0.7...v0.0.8) (2022-03-18)

* Created a standalone function for `Reachability` and updated `openCheckUrl` to call `getReachability`.

### [0.0.7](https://github.com/tru-ID/tru-sdk-web/compare/v0.0.5...v0.0.7) (2022-02-16)

* re-publish because of missing `/dist` folder in `0.0.6`

### [0.0.6](https://github.com/tru-ID/tru-sdk-web/compare/v0.0.5...v0.0.6) (2022-01-06)

* re-publish using proper tags
* Updated readme to no longer use `canary`

### [0.0.5](https://github.com/tru-ID/tru-sdk-web/compare/v0.0.4...v0.0.5) (2021-06-10)

* re-publish using proper tags

### 0.0.4 (2021-06-10)


### Bug Fixes

* file extensions for cjs & mjs
* unresolved promise when image loads successfully
* use || instead of ??

### [0.0.3](https://github.com/tru-ID/tru-sdk-web/compare/v0.0.2-canary.1...v0.0.3) (2021-06-07)

- Fix default export

### 0.0.2-canary.0 (2021-06-01)

- Automatically check device coverage to determine if it's not a mobile IP address or not a supported mobile network.
- Added option to bypass this check with `checkDeviceCoverage: false`
- Updated README and packages

### Bug Fixes

- file extensions for cjs & mjs ([5bba996](https://github.com/tru-ID/tru-sdk-web/commit/5bba99671c676cb4710b0c3402f0d411bfc6b860))
- use || instead of ?? ([031057c](https://github.com/tru-ID/tru-sdk-web/commit/031057c198aab39ba0f7bcedca2d23b1fabfd37a))

### [0.0.1-canary.2](https://gitlab.com/4auth/devx/tru-sdk-web/compare/v0.0.1-canary.1...v0.0.1-canary.2) (2021-02-17)

### Bug Fixes

- file extensions for cjs & mjs ([5bba996](https://github.com/4auth/devx/tru-sdk-web/commit/5bba99671c676cb4710b0c3402f0d411bfc6b860))
- use || instead of ?? ([031057c](https://github.com/4auth/devx/tru-sdk-web/commit/031057c198aab39ba0f7bcedca2d23b1fabfd37a))

### [0.0.1-canary.1](https://github.com/4auth/devx/tru-sdk-web/compare/v0.0.1-canary.0...v0.0.1-canary.1) (2021-02-02)

- Updated README with installation and usage instructions

### 0.0.1-canary.0 (2021-02-01)

- Supports installation via CDN and NPM
- Exposes `openCheckUrl(checkUrl, config)`


tests/index.test.js
require('jest-fetch-mock')

const nock = require('nock')
const truID = require('../dist/tru-id-sdk.js')

const baseUrl = 'https://eu.api.tru.id'

const testUrl = `${baseUrl}/phone_check/v0.1/checks/12345/redirect`

describe('WEB SDK', () => {
  afterAll(nock.restore)
  afterEach(nock.cleanAll)

  test('throw an error if no url is passed', async () => {
    expect(truID.openCheckUrl()).rejects.toThrow()
  })

  test('getReachability is successful', async () => {
    const coverageResponse = {
      country_code: "GB",
      network_id: "12453",
      network_name: "EE UK",
      products: [{product_id: "PCK", product_name: "Phone Check"}, {product_id: "SCK", product_name: "Sim Check"}, {product_id: "SUK", product_name: "Subscriber Check"}],
      _links: {self: {href: "https://eu.api.tru.id/public/coverage/v0.1/device_ip"}},
    };

    const coverage = nock(baseUrl)
      .get('/public/coverage/v0.1/device_ip')
      .reply(200, coverageResponse)

    const res = await truID.getReachability(testUrl)
    expect(coverage.isDone()).toBe(true)
    expect(res).toHaveProperty('reachable', true)
    expect(res.body).toHaveProperty('country_code', coverageResponse.country_code)
  })

  test('getReachability failed with error "Not Mobile IP"', async () => {
    const coverageResponse = {
      type: "https://developer.tru.id/docs/reference/api-errors#not_mobile_ip",
      title: "Precondition Failed",
      detail: "Not a mobile IP"
    };

    const coverage = nock(baseUrl)
    .get('/public/coverage/v0.1/device_ip')
    .reply(412, coverageResponse)

    const res = await truID.getReachability(testUrl)
    expect(coverage.isDone()).toBe(true)
    expect(res).toHaveProperty('reachable', false)
    expect(res.body).toHaveProperty('detail', coverageResponse.detail)
  })

  test('phone check is successful', async () => {
    const coverage = nock(baseUrl)
      .get('/public/coverage/v0.1/device_ip')
      .reply(200, {})
    const redirect = nock(baseUrl)
      .get('/phone_check/v0.1/checks/12345/redirect')
      .replyWithFile(200, 'tests/logo.svg', {
        'Content-Type': 'image/svg+xml',
      })
    await truID.openCheckUrl(testUrl)
    expect(coverage.isDone()).toBe(true)
    expect(redirect.isDone()).toBe(true)
  })

  test('should not to call the checkUrl when MNO not supported', async () => {
    const coverage = nock(baseUrl)
      .get('/public/coverage/v0.1/device_ip')
      .reply(400, {
        detail: 'MNO not supported',
      })
    const redirect = nock(baseUrl)
      .get('/phone_check/v0.1/checks/12345/redirect')
      .replyWithFile(200, 'tests/logo.svg', {
        'Content-Type': 'image/svg+xml',
      })
    await expect(truID.openCheckUrl(testUrl)).rejects.toHaveProperty(
      'code',
      truID.DeviceCoverageErrors.UnsupportedMNO,
    )
    expect(coverage.isDone()).toBe(true)
    // check url never called
    expect(redirect.isDone()).toBe(false)
  })

  test('should not call the checkUrl when not mobile IP ', async () => {
    const coverage = nock(baseUrl)
      .get('/public/coverage/v0.1/device_ip')
      .reply(412, {
        detail: 'Not mobile IP',
      })
    const redirect = nock(baseUrl)
      .get('/phone_check/v0.1/checks/12345/redirect')
      .replyWithFile(200, 'tests/logo.svg', {
        'Content-Type': 'image/svg+xml',
      })
    await expect(truID.openCheckUrl(testUrl)).rejects.toHaveProperty(
      'code',
      truID.DeviceCoverageErrors.NotMobileIP,
    )
    expect(coverage.isDone()).toBe(true)
    // check url never called
    expect(redirect.isDone()).toBe(false)
  })

  test('should not proceed when "checkDeviceCoverage: false" is passed', async () => {
    const coverage = nock(baseUrl)
      .get('/public/coverage/v0.1/device_ip')
      .reply(412, {
        detail: 'Not mobile IP',
      })
    const redirect = nock(baseUrl)
      .get('/phone_check/v0.1/checks/12345/redirect')
      .replyWithFile(200, 'tests/logo.svg', {
        'Content-Type': 'image/svg+xml',
      })
    await truID.openCheckUrl(testUrl, {
      checkDeviceCoverage: false,
    })
    // device coverage url is never called
    expect(coverage.isDone()).toBe(false)
    // check url is called
    expect(redirect.isDone()).toBe(true)
  })
})


tests/index.test.mjs
import truID from '../dist/tru-id-sdk'

const testUrl = 'https://example.com'

test('iframe is added to the DOM', () => {
    truID.openCheckUrl(testUrl)
    expect(document.body.children[0].tagName).toBe('IFRAME')
})

test('iframe `src` is set', () => {
    const expectedSrc = truID._createIFrameSrc(testUrl)
    truID.openCheckUrl(testUrl)
    expect(document.body.children[0].src).toBe(expectedSrc)
})

src/index.js
export const DeviceCoverageErrors = {
  UnsupportedMNO: 400,
  NotMobileIP: 412,
}

export function _createIFrameSrc(checkUrl, config = { debug: false }) {
  const html = `
		<body><script>
		const debug = ${config.debug}
		const img = new Image()
		img.style.height = 0
		img.style.width = 0
		img.setAttribute('referrerpolicy', 'no-referrer')
		img.src = "${checkUrl}"
	
		function handleEndEvent() {
			if(debug) {
				console.log('image loaded')
			}
			window.parent.postMessage({ error: false}, "${window.origin}")
		}

        function handleLoadError() {
            if(debug) {
                console.log('error loading image')
            }
            window.parent.postMessage({ error: true }, "${window.origin}")
        }
	
		img.onload = handleEndEvent
		img.onerror = handleLoadError
		document.body.appendChild(img)
		</script></body>
		`
  const src = `data:text/html;charset=utf-8,${encodeURI(html)}`
  return src
}

/**
 * Opens the provided `apiBaseUrl` in order to perform the Reachability check.
 *
 * A Reachability check is a test if the current device's active data connection is one of a mobile carrier, and that this carrier is supported by tru.ID's Phone, Subscriber, and SIM checks.
 *
 * @param {String} apiBaseUrl The tru.ID `base_url` used to communicate with the tru.ID APIs. For example: `https://eu.api.tru.id`.
 * 
 * @return { reachable: Boolean, body: { network_id: String, network_name: String, country_code: String, products: [] }, status: Integer }
 */
export async function getReachability(apiBaseUrl) {
  if (!apiBaseUrl) {
    throw new Error('apiBaseUrl is required')
  }

  const url = new URL(apiBaseUrl)
  const baseUrl = url.origin

  const apiResponse = await fetch(`${baseUrl}/public/coverage/v0.1/device_ip`)
  const responseBody = await apiResponse.json()

  if (apiResponse.status === 200) {
    return { reachable: true, body: responseBody, status: apiResponse.status }
  }

  return { reachable: false, body: responseBody, status: apiResponse.status }
}

/**
 * Opens the provide `checkUrl` in order to perform the PhoneCheck.
 *
 * A PhoneCheck is a test if the current device has authenticated with the mobile carrier using the expected IMSI (SIM Card) and MSISDN (Phone Number).
 *
 * @param {String} checkUrl The tru.ID `check_url` to be requested over the mobile data network as part of the verification workflow.
 * @param {Object} customConfig
 * @param {String} customConfig.checkMethod When set to `image` the library will load the `check_url` using a zero pixel image loaded within an <iframe>. When set to `window` the check_url will briefly launch a window that will then be closed.
 * @param {Number} customConfig.windowCloseTimeout The timeout period in milliseconds where the window is closed if `customConfig.checkMethod` is set to `window`. Defaults to 3000 milliseconds.
 * @param {Boolean} customConfig.checkDeviceCoverage It will not try to check the device IP against the coverage API.
 * @param {Boolean} customConfig.debug It will console log more debug info.
 * @param {String} customConfig.version Based on the PhoneCheck API version to determine the method of flow.
 */
export async function openCheckUrl(checkUrl, customConfig) {
  const defaultConfig = {
    checkMethod: 'image',
    debug: false,
    checkDeviceCoverage: true,
    windowCloseTimeout: 3000,
    version: 'v0.2',
  }

  const config = Object.assign(defaultConfig, customConfig)

  function log(...args) {
    if (config.debug) {
      console.log('tru.ID:sdk-web ', ...args)
    }
  }

  if (!checkUrl) {
    throw new Error('checkUrl is required')
  }

  // Check device coverage
  // unless the user is passing checkDeviceCoverage:false to skip the check
  if (config.checkDeviceCoverage) {
    const coverage = await getReachability(checkUrl)

    if (coverage.status !== 200) {
      const err = new Error(coverage.body)
      err.body = coverage.body
      err.code = coverage.status

      throw err
    }
  }

  return new Promise((resolve, reject) => {
    if (config.version === 'v0.2') {
      window.location = checkUrl

      return
    }

    if (config.checkMethod === 'image') {
      const iframe = document.createElement('iframe')
      iframe.style.display = 'none'
      iframe.setAttribute('sandbox', 'allow-same-origin allow-scripts')
      iframe.src = _createIFrameSrc(checkUrl)

      const handleIFrameMessage = (event) => {
        log(`${JSON.stringify(event)}`)
        window.removeEventListener('message', handleIFrameMessage)

        if (event.error) {
          reject(new Error('Error loading invisible image'))
        } else {
          document.body.removeChild(iframe)
          resolve()
        }
      }
      window.addEventListener('message', handleIFrameMessage)

      document.body.appendChild(iframe)
    } else {
      const win = window.open(checkUrl)
      setTimeout(() => {
        win.close()
        resolve()
      }, config.windowCloseTimeout)
    }
  })
}

export default {
  DeviceCoverageErrors,
  openCheckUrl,
  getReachability,
}
